airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/util/RecordMigrations.java

```diff
@@ -0,0 +1,267 @@
+        isObjectSchema = "object".equals(typeNode.asText());
+      }
+    } else {
+      // If the schema doesn't declare a type at all (which is bad practice, but let's handle it anyway)
+      // Then check for a properties entry, and assume that this is an object if it's present
+      isObjectSchema = schema.hasNonNull("properties");
+    }
+
+    if (!isObjectSchema) {
+      // If it's not supposed to be an object, then we can't do anything here.
+      // Return the data without modification.
+      return new MigratedNode(data, false);
+    } else {
+      // If the schema _is_ for an object, then recurse into each field
+      ObjectNode mutatedData = (ObjectNode) Jsons.emptyObject();
+      JsonNode propertiesNode = schema.get("properties");
+
+      Iterator<Entry<String, JsonNode>> dataFields = data.fields();
+      boolean matchedSchema = true;
+      while (dataFields.hasNext()) {
+        Entry<String, JsonNode> field = dataFields.next();
+        String key = field.getKey();
+        JsonNode value = field.getValue();
+        if (propertiesNode != null && propertiesNode.hasNonNull(key)) {
+          // If we have a schema for this property, mutate the value
+          JsonNode subschema = propertiesNode.get(key);
+          MigratedNode migratedNode = mutateDataNode(validator, schemaMatcher, transformer, value, subschema);
+          mutatedData.set(key, migratedNode.node);
+          if (!migratedNode.matchedSchema) {
+            matchedSchema = false;
+          }
+        } else {
+          // Else it's an additional property - we _could_ check additionalProperties,
+          // but that's annoying. We don't actually respect that in destinations/normalization anyway.
+          mutatedData.set(key, value);
+        }
+      }
+
+      return new MigratedNode(mutatedData, matchedSchema);
+    }
+  }
+
+  /**
+   * Much like objects, arrays must be recursively mutated.
+   */
+  private static MigratedNode mutateArrayNode(
+                                              JsonSchemaValidator validator,
+                                              Function<JsonNode, Boolean> schemaMatcher,
+                                              Transformer transformer,
+                                              JsonNode data,
+                                              JsonNode schema) {
+    // Similar to objects, we first check whether this is even supposed to be an array.
+    boolean isArraySchema;
+    if (schema.hasNonNull(REF_KEY)) {
+      // If the schema uses a reference type, then it's not an array schema.
+      isArraySchema = false;
+    } else if (schema.hasNonNull(TYPE_KEY)) {
+      // If the schema declares {type: array} or {type: [..., array, ...]}
+      // Then this is an array schema
+      JsonNode typeNode = schema.get(TYPE_KEY);
+      if (typeNode.isArray()) {
+        isArraySchema = false;
+        for (JsonNode typeItem : typeNode) {
+          if ("array".equals(typeItem.asText())) {
```

Comment:
Maybe it would be good to have these "arrays", "properties" etc also stored in a static variables. Could be pretty useful if we will have to change it in the future - so will be able to change it just in a single place

Reply:
added these constants to JsonSchemaReferenceTypes